Userland/Libraries/LibGfx/ImageFormats/TGALoader.cpp

```diff
@@ -278,30 +132,43 @@ ErrorOr<ImageFrameDescriptor> TGAImageDecoderPlugin::frame(size_t index, Optiona
     if (x_origin != 0 && x_origin != width)
         return Error::from_string_literal("TGAImageDecoderPlugin: Can only handle X origin which is 0 or the entire width");
 
+    VERIFY((bits_per_pixel % 8) == 0);
+    auto pixel_buffer = TRY(ByteBuffer::create_zeroed(bits_per_pixel / 8));
+
     switch (data_type) {
     case TGADataType::UncompressedRGB: {
         for (int row = 0; row < height; ++row) {
             for (int col = 0; col < width; ++col) {
-                auto pixel = m_context->reader->read_pixel(bits_per_pixel);
+                TRY(m_context->stream.read_until_filled(pixel_buffer.bytes()));
+                auto pixel_data = TGAPixelData(pixel_buffer.bytes());
                 auto actual_row = row;
                 if (y_origin < height)
                     actual_row = height - 1 - row;
                 auto actual_col = col;
                 if (x_origin > width)
                     actual_col = width - 1 - col;
-                bitmap->scanline(actual_row)[actual_col] = pixel.data;
+                bitmap->scanline(actual_row)[actual_col] = pixel_data.raw;
             }
         }
         break;
     }
+
     case TGADataType::RunLengthEncodedRGB: {
+        auto pixel_packet_type_buffer = TRY(ByteBuffer::create_zeroed(sizeof(u8)));
         size_t pixel_index = 0;
         size_t pixel_count = height * width;
         while (pixel_index < pixel_count) {
-            auto packet_type = m_context->reader->read_packet_type();
-            VERIFY(packet_type.pixels_count > 0);
-            TGAPixel pixel = m_context->reader->read_pixel(bits_per_pixel);
-            auto max_pixel_index = min(pixel_index + packet_type.pixels_count, pixel_count);
+            TRY(m_context->stream.read_until_filled(pixel_packet_type_buffer.bytes()));
```

Comment:
```cpp
auto const pixel_packet_type = TRY(m_context->stream.read_value<u8>());
```

And it would be nice to keep the logic to read the packet type in a function